crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+				// check provided attribute has the correct type
+				if *schema_attribute_type == provided_attribute.type_id() {
+					ensure!(provided_attribute.len() <= T::MaxAttributeLength::get() as usize, Error::<T>::MaxAttributeLength);
+					Ok(provided_attribute.clone())
+				} else {
+					Err(Error::<T>::SchemaMismatch)
+				}
+			}).collect::<Result<_, Error<T>>>()?;
+
+			// Create the token, update ownership, and bookkeeping
+			if let Some(royalties_schedule) = royalties_schedule {
+				ensure!(royalties_schedule.validate(), Error::<T>::RoyaltiesOvercommitment);
+				<TokenRoyalties<T>>::insert(&collection_id, token_id, royalties_schedule);
+			}
+			<Tokens<T>>::insert(&collection_id, token_id, token);
+			<NextTokenId<T>>::insert(&collection_id, next_token_id);
+			<TokenIssuance<T>>::mutate(&collection_id, |i| *i += One::one());
+			<TokenOwner<T>>::insert(&collection_id, token_id, owner.clone());
+			<CollectedTokens<T>>::append(&collection_id, owner.clone(), token_id);
+
+			Self::deposit_event(RawEvent::CreateToken(collection_id, token_id, owner));
+
+			Ok(())
+		}
+
+		/// Transfer ownership of an NFT
+		/// Caller must be the token owner
+		#[weight = T::WeightInfo::transfer()]
+		fn transfer(origin, collection_id: CollectionId, token_id: T::TokenId, new_owner: T::AccountId) {
+			let origin = ensure_signed(origin)?;
+
+			ensure!(CollectionSchema::contains_key(&collection_id), Error::<T>::NoCollection);
+			ensure!(<Tokens<T>>::contains_key(&collection_id, token_id), Error::<T>::NoToken);
+
+			let current_owner = Self::token_owner(&collection_id, token_id);
+			ensure!(current_owner == origin, Error::<T>::NoPermission);
+
+			ensure!(!<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::TokenListingProtection);
+
+			Self::transfer_ownership(&collection_id, token_id, &current_owner, &new_owner);
+			Self::deposit_event(RawEvent::Transfer(collection_id, token_id, new_owner));
+		}
+
+		/// Burn an NFT 🔥
+		/// Caller must be the token owner
+		#[weight = T::WeightInfo::burn()]
+		fn burn(origin, collection_id: CollectionId, token_id: T::TokenId) {
+			let origin = ensure_signed(origin)?;
+
+			ensure!(CollectionSchema::contains_key(&collection_id), Error::<T>::NoCollection);
+			ensure!(<Tokens<T>>::contains_key(&collection_id, token_id), Error::<T>::NoToken);
+
+			let current_owner = Self::token_owner(&collection_id, token_id);
+			ensure!(current_owner == origin, Error::<T>::NoPermission);
+
+			ensure!(!<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::TokenListingProtection);
+
+			// Update token ownership
+			<CollectedTokens<T>>::mutate(&collection_id, current_owner, |tokens| {
+				tokens.retain(|t| t != &token_id)
+			});
+			<TokenOwner<T>>::take(&collection_id, token_id);
+			<Tokens<T>>::take(&collection_id, token_id);
+
+			// Will not overflow, cannot exceed the amount issued qed.
+			let tokens_burnt = Self::tokens_burnt(&collection_id).checked_add(&One::one()).unwrap();
```

Comment:
Unwrap can potentially crash the network. Should avoid at all costs.

Reply:
this has a 'qed' on the line above, same as substrate practice

Reply:
I see. I would still suggest `unwrap_or` or similar instead of playing with a hot burning ball. Who knows how the program would evolve next time someone is touching this code and if they make a mistake. The side effect of a mistake in a production level code shouldn't be a crash.